pilot/pkg/xds/eds.go

```diff
@@ -462,11 +474,26 @@ func (eds *EdsGenerator) buildEndpoints(proxy *model.Proxy,
 			}
 		}
 		builder := NewEndpointBuilder(clusterName, proxy, req.Push)
-		if marshalledEndpoint, f := eds.Server.Cache.Get(builder); f && !features.EnableUnsafeAssertions {
-			// We skip cache if assertions are enabled, so that the cache will assert our eviction logic is correct
-			resources = append(resources, marshalledEndpoint)
-			cached++
-		} else {
+
+		// We skip cache if assertions are enabled, so that the cache will assert our eviction logic is correct
+		if !features.EnableUnsafeAssertions {
+			var cachedEndpoint *discovery.Resource
+			if useFullCache {
+				// If no PA changed
+				cachedEndpoint = eds.Server.Cache.Get(&builder)
+			} else {
+				// If PA has changed but the eds cache has already been refreshed in this period.
+				cachedEndpoint = eds.Server.Cache.GetWithRequest(&builder, req)
```

Comment:
Hmm... thinking about this more I am a bit worried this may not be sound. Consider we have 3 proxies. Timeline:


* Push 0, store in cache with PA v0. Cache={PAv0}
* Push 1 starts with PA v1
* Proxy 1 generates and stores in cache. Cache={PAv1}
* Push 2 starts with PA v2
* Proxy 1 generates and stores in cache. Cache={PAv2}
* Proxy 2 finally gets to generating, Its request has push context with PAv1 still. It loads PAv2 though

Although I think this would be broken even before this change... let me think about it a bit more, maybe this is impossible

Reply:
> It loads PAv2 though


We have this problem before, and it always this.  Since in istio we use eventual consistent model, I cannot think of a concrete sideeffect

Reply:
I think there is still an issue but not what i mentioned

1. Push 0, store in cache with PA v0. Cache={PAv0}
2. Push 1 starts with PA v1
3. Proxy 1 generates and stores in cache. Cache={PAv1}
4. Push 2 starts with PA v2
5. Proxy 1 generates and stores in cache. Cache={PAv2}. ASSERTION FAILS HERE

The assertion logic checks that we never write two same keys with different values. Here we violate this. It was previously ok because we would have evicted the first key when PA v2 is created?

Reply:
Good catch, it needs to handle this special case. I will think about it